### ISSUE TITLE
Inventory adjustments frailties

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#1812]
_Imported from trac ticket 1812.
Created by **kbower**
Opened in trac: Jul 16, 2012, 4:54:10 PM
Last modified in trac: Oct 29, 2012, 1:46:04 PM_
***
Inventory adjustments are not saved through the existing framework inventory/save controller.
All framework mechanisms such as restricted attributes and checks for meantime changes to the inventory record since the "get" are being bypassed.

This needs to be refactored to use the "get/save" mechanism instead of being bypassed in the controller. (This will have another positive side effect of moving this object logic centrally to Inventory.)

Further, there is apparently no safety mechanism for duplication of http request.  Fixing the above may naturally solve this problem, possibly.

***
* kbower attached [None](https://trac.retailarchitects.com/trac/attachment/ticket/1812/None) on Oct 3, 2012, 11:07:28 AM
* kbower attached [None-1](https://trac.retailarchitects.com/trac/attachment/ticket/1812/None-1) on Oct 3, 2012, 11:07:29 AM

**Comment by kbower** [comment]
_**kbower** trac comment on Jul 16, 2012, 4:56:05 PM:_

Rob, let me know your intention regarding this ticket.  You plan to take ownership or should it be assigned elsewhere?

**Comment by kbower** [comment]
_**rneville** trac comment on Jul 16, 2012, 5:23:34 PM:_

There are issues with this that are not straightforward trying to ram it into the typical db entity save

This is not a blocker as no one is using it yet.

**Comment by kbower** [comment]
_**kbower** trac comment on Jul 19, 2012, 5:18:28 PM:_

So really, they block each other, and its a deadlock with [#1818](1818) ;)

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 29, 2012, 8:43:27 AM:_

(Changeset [[3232]](../commit/7d91df5c91fca7c70c7d0c9ab5afe0961aa99b5d))
 * Support V12 Inventory (closes [#1818](1818)).
   * Inventory class definition is set based on gers version and metadata dynamically sent to client
   * Note that an `InventoryCost` class exists in both version, but is a _very_ different record type for each version
   * add global `cfg.gers_version` as the gers version (set in app.ini)
   * some apparently unused (by morris) costing tables are not implemented 
 * Save Inventory adjustments through framework save (fixes #1812), ensuring record locking, restricted_attributes, and detection of concurrent change 
 * fixes [#1942](1942) (overagereasonid not being recorded for new inventory adjustment)
 * fixes [#1943](1943) (Overage inventory adjustments need to record site and location in 'new' transaction fields instead of 'old')
 * verify that 'readonly' columns are not passed from client for new records, unless marked as 'readonly':'persistent', which means they are only readonly after the record is persisted to the database
 * Add session.perform_after_flush() method for registering function invocation the must wait until flush is over
 * During save(), if the transferObject itself is deleted, mark the object `instance._dead_object_ = True` so that the refresh from the database knows not to raise ExpectedRecordNotFound
 * record both RSequence and Sequence in metadata.sequences (and create them during database upgrade)
 * remove flex code of InventoryTran query and make it dynaform (with many more available search fields and columns)

**Comment by kbower** [comment]
_**kbower** trac status change on Sep 29, 2012, 8:43:27 AM:_ **closed**

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 3, 2012, 11:07:28 AM:_

Incidentally, I went down the road of making "New Inventory..." support the idea of "merging", i.e., make the system see that it was existing inventory and update the existing record instead of complaining the record already existed.  I abandoned that for a several reasons:
 * I had to horribly hack many places in an ugly way to make this exception work.  (As it is now, it is just the normal framework inventory/save)
 * The Cost field is confusing in the event your inventory record already exists, since you think you are entering a cost, but it is doomed to be discarded.  Then why is it there?
 * After the save took place, the New Qty reverted to the /real /quantity which is just too confusing and unprofessional... "Wait, I said qty 4, why is it suddenly qty 20!?"
 * It was unclear whether the user realized the inventory was already there.  What does "New Qty" mean?  Does it mean the new current qty or just the delta?  How the system deals with the input depends on the answer and the answer is impossible to determine unless you know whether or not the user is aware the record already exists.
 * In the end, the last reason listed is the most important.  The user must communicate they know what is going on by correctly choosing "New inventory..." or "Adjust inventory..."  Conservative, no surprises, no ambiguities.

[[Image(None)]]

[[Image(None-1)]]
